### PR TITLE
fix: Git tags empty

### PR DESCRIPTION
**Pull request by Jayrgo**
if `skipVersionFile` or `skipCommit` is set, then `checkout` must be used `with: fetch-depth: 0`

```yaml
- uses: actions/checkout@v2
  with:
    fetch-depth: 0
- name: Conventional Changelog Action
  uses: TriPSs/conventional-changelog-action@v3
  with:
    skip-version-file: 'true'
    skip-commit: 'true'
```

### CODE REVIEW

File: src/helpers/git.js [hunk 1]
```diff
@@ -97,7 +97,7 @@ module.exports = new (class Git {
    * @return {Promise<>}
    */
   pull = () => (
-    this.exec(`pull --unshallow ${core.getInput('git-pull-method')}`)
+    this.exec(`pull ${core.getInput('git-pull-method')}`)
```
Comment (TriPSs) [inline src/helpers/git.js:100]: Is there a reason why you removed the unshallow? That should also make sure all the git tags are retrieved.

File: src/version/git.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ module.exports = new (class Git extends BaseVersioning {
       gitSemverTags({
         tagPrefix,
       }, (err, tags) => {
-        const currentVersion = tags.shift().replace(tagPrefix, '')
+        const currentVersion = tags.length > 0 ? tags.shift().replace(tagPrefix, '') : '0.1.0'
```
Comment (TriPSs) [inline src/version/git.js:16]: Maybe make the initial version configurable?

File: src/helpers/git.js (hunk 1 above)
Comment (Jayrgo) [reply src/helpers/git.js:100]: i have created a test repository with a workflow that runs `git log --decorate --no-color`

```yaml
- name: Checkout
  uses: actions/checkout@v2
  with:
    fetch-depth: 0
- name: Pull
  run: |
    git pull --unshallow --ff-only
```
--> `fatal: --unshallow on a complete repository does not make sense`
```yaml
steps:
  - name: Checkout
    uses: actions/checkout@v2
  - name: Pull
    run: |
      git pull --unshallow --ff-only
```
--> no tags available in log
... `commit f5d3880d5518d39253e3a0f59c07bc13c824a762` ...

```yaml
steps:
  - name: Checkout
    uses: actions/checkout@v2
    with:
      fetch-depth: 0
  - name: Pull
    run: |
      git pull --ff-only
```
--> tags available in log 
... `commit f5d3880d5518d39253e3a0f59c07bc13c824a762 (tag: 0.1.0)` ...

Comment (TriPSs) [reply src/helpers/git.js:100]: Thanks for the explanation, personally I would not remove it as doing so and requiring uses to update their checkout action will result in a breaking change. Think its better for this pr to only do the configurable fallback version and pheraps change this in v4.

Comment (Jayrgo) [reply src/helpers/git.js]: found another solution:
`git pull --unshallow --tags --ff-only` --> ... `commit f5d3880d5518d39253e3a0f59c07bc13c824a762 (tag: 0.1.0)` ...

File: package.json [hunk 3]
```diff
@@ -26,11 +26,11 @@
   "dependencies": {
     "@actions/core": "1.2.4",
     "@actions/exec": "1.0.4",
+    "@iarna/toml": "^2.2.5",
     "conventional-changelog": "3.1.21",
     "conventional-recommended-bump": "6.0.9",
     "git-semver-tags": "4.0.0",
     "object-path": "^0.11.4",
-    "toml": "^3.0.0",
```
Comment (TriPSs) [inline package.json:33]: There a reason for the switch?

Comment (Jayrgo) [reply package.json:33]: yes
this dependency has no `stringify`
this is used if the version can't be read from the file and we need to write a valid toml file

File: src/helpers/bumpVersion.js [hunk 4]
```diff
@@ -1,3 +1,6 @@
+const core = require('@actions/core')
+const semverValid = require('semver').valid
```
Comment (TriPSs) [inline src/helpers/bumpVersion.js:2]: Just noticed that `semver` is not in our package.json, could we add it there?

Comment (Jayrgo) [reply src/helpers/bumpVersion.js:2]: done in https://github.com/TriPSs/conventional-changelog-action/pull/50/commits/283100b0c8b51e728cf3c1553169172ae432b703

**Comment by TriPSs** [comment]
When running `act -j test-json-empty -P ubuntu-latest=nektos/act-environments-ubuntu:18.04 -s github_token=fake-token` it errors:
```
(node:13) UnhandledPromiseRejectionWarning: SyntaxError: Unexpected end of JSON input
|     at JSON.parse (<anonymous>)
|     at Json.bump (/github/workspace/src/version/json.js:17:30)
|     at handleVersioningByExtension (/github/workspace/src/index.js:20:20)
|     at /github/workspace/src/index.js:91:18
|     at Array.map (<anonymous>)
|     at /github/workspace/src/index.js:88:52
|     at /github/workspace/node_modules/conventional-recommended-bump/index.js:93:11
|     at ConcatStream.<anonymous> (/github/workspace/node_modules/concat-stream/index.js:37:43)
|     at ConcatStream.emit (events.js:323:22)
|     at finishMaybe (/github/workspace/node_modules/readable-stream/lib/_stream_writable.js:624:14)
| (node:13) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
| (node:13) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
Same happend with `-new`, `toml` / `yaml` worked as expected.

Als what maybe it's handy to add a log when the fallback version is used so users are not confused where the version came from.

Everything looks good so after the fix for the new / empty JSON file it should be ready to go, thanks for all the work!

**Comment by Jayrgo** [comment]
> When running `act -j test-json-[empty/new] -P ubuntu-latest=nektos/act-environments-ubuntu:18.04 -s github_token=fake-token` it errors:

yeah, this is correct, because an empty Json file/string is not a valid Json.
there are several solutions:
- try/catch and use fallback https://github.com/TriPSs/conventional-changelog-action/pull/50/commits/97f1bb3543e6f2480ef3e699fc695ecb8b3f881b
- display this or another error message and abort

> Als what maybe it's handy to add a log when the fallback version is used so users are not confused where the version came from.

https://github.com/TriPSs/conventional-changelog-action/pull/50/commits/9a324dbd51d0d32c1b9df1a291e14cc20a5bbaff

File: src/helpers/bumpVersion.js [hunk 5]
```diff
@@ -6,23 +9,40 @@
...
+    let version = semverValid(core.getInput('fallback-version'))
+
+    if (version) {
+      [major, minor, patch] = version.split('.')
+    } else {
+      // default
+      major = 0
+      minor = 1
+      patch = 0
+    }
```
Comment (TriPSs) [inline src/helpers/bumpVersion.js:42]:
```suggestion
    }
     
     core.info(`The version could not be detected, Using fallback version '${major}.${minor}.${patch}'.`)
```

File: src/helpers/bumpVersion.js [hunk 6]
```diff
@@ -6,23 +9,40 @@
...
   }
 
+  core.info(`The version could not be detected, the new version is '${major}.${minor}.${patch}'.`)
```
Comment (TriPSs) [inline src/helpers/bumpVersion.js:45]: Should this not be inside the `else`? As it's logged now every time, see my suggestion

```suggestion
```

**Comment by TriPSs** [comment]
Everything looks good! Thanks for the PR!

**Pull request merged**